File: src/lib/services/sleep-inhibit.ts
import { invoke } from '@tauri-apps/api/core';

export async function withSleepInhibit<T>(
  reason: string,
  fn: () => Promise<T>
): Promise<T> {
  let token: number | null = null;

  try {
    token = await invoke<number>('acquire_sleep_inhibit', { reason });
  } catch (error) {
    console.error('Failed to acquire sleep inhibition:', error);
  }

  try {
    return await fn();
  } finally {
    if (token !== null) {
      try {
        await invoke('release_sleep_inhibit', { token });
      } catch (error) {
        console.error('Failed to release sleep inhibition:', error);
      }
    }
  }
}

File: src-tauri/src/tools/power/sleep_inhibit.rs
#[tauri::command]
pub(crate) async fn acquire_sleep_inhibit(reason: String) -> Result<u64, String> {
    crate::shared::sleep_inhibit::acquire_sleep_inhibit(reason)
}

#[tauri::command]
pub(crate) async fn release_sleep_inhibit(token: u64) -> Result<(), String> {
    crate::shared::sleep_inhibit::release_sleep_inhibit(token)
}

